Add 'json escape' option to unicode escape <, >, and &

g-k opened the pull request:
refs: #3268

Commit:
Add 'json escape' option to unicode escape <, >, and &

dougwilson reviewed (changes requested):
So I am reviewing this with the assumption that #3268 will result in the desire to add this feature, but you can wait on that discussion to fix the issue if you like :)

I would expect that `res.jsonp` would respect this option, if you can fix that (i.e. I would expect anywhere in Express that respects `json replacer`, `json spaces`, etc. to also honor this new `json escape`).

Commit:
Add 'escape json' option for jsonp responses

Commit:
use single replace call like DavidTPate recommended for hoek

could use the char code version too

g-k commented:
Added the setting to jsonp and git grep doesn't show the json options used elsewhere.

Updated to use a single String.replace call that a similar change to hapijs/hoek showed as ~3x faster on the twitter.json test data.

Commit:
fix coverage

dougwilson commented:
Unless there are any objections, I'm suggesting that this land in the 4.16.0 release.

dougwilson commented on lib/response.js:
```diff
@@ -248,6 +248,18 @@ res.json = function json(obj) {
 
+  // escape characters that can cause old browsers to mistake json for html
+  if (app.get('json escape')) {
+      body = body.replace(/[<>&]/g, function (match) {
```
Please hoist regular expressions to the top of the file.

dougwilson replied:
And also since the function close not close over anything, please hoist that as well.

dougwilson commented on lib/response.js:
```diff
@@ -311,11 +323,28 @@ res.jsonp = function jsonp(obj) {
 
+    // escape characters that can cause old browsers to mistake json for html
+    if (app.get('json escape')) {
+      body = body.replace(/[<>&\u2028\u2029]/g, function (match) {
```
Same comments as the other replace: hoist regular expression and function.

dougwilson commented on test/res.json.js:
```diff
@@ -144,6 +144,28 @@ describe('res', function(){
+    describe('"json escape" setting', function(){
+      it('should be undefined by default', function(){
+        var app = express();
+        assert(undefined === app.get('json escape'));
```
I would really just say it's off by default and run through a response to validate it doesn't get escaped. That would help in case a change accidentally unconditionally escapes it.

dougwilson commented on test/res.jsonp.js:
```diff
@@ -112,6 +112,21 @@ describe('res', function(){
       .expect(200, '{"str":"\u2028 \u2029 woot"}', done);
     });
 
+    it('should escape <, >, and & when "json escape" is true', function(done){
```
I think this is missing a default off test, like the res.json one :)

dougwilson reviewed (changes requested):
Final comments added. Ideally all changed code adheres to the StandardJS style, but I can alter your changes on landing if you don't want to change that up.

Commit:
response: hoist regexps and escape fns

Commit:
response: try to get patch standard@10.0.2 compliant

Commit:
test: test res.json and res.jsonp don't unicode escape by default

Commit:
test: try to get new tests standardjs compliant

g-k commented:
Thanks @dougwilson! I think I got all of the requested changes, but please modify the PR as necessary to get it in shape too.
